### ISSUE TITLE
Phase 7.2 — Task distribution sampler: TaskSampler, curriculum scheduling, and difficulty-aware sampling

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#262]
## Phase 7.2 — Task Distribution Sampler

**Phase 7.1** introduced `ReptileMetaLearner` with an inner-loop adaptation loop. For meta-learning to work well, the **distribution of tasks** presented to the meta-learner matters as much as the algorithm itself. This sub-phase introduces a structured task sampling system with curriculum support.

---

## Why task sampling matters

Reptile (and meta-learning in general) learns a prior over tasks. The quality of that prior depends on:

| Factor | Poor sampling | Good sampling |
|---|---|---|
| Task diversity | Learner overfits to a few tasks | Learner generalises broadly |
| Task difficulty | All hard tasks → slow convergence | Curriculum: easy → hard → plateau |
| Task balance | Skewed distribution → biased prior | Uniform / importance-weighted |
| Freshness | Stale tasks dominate | Recency-weighted or shuffled |

A `TaskSampler` abstracts this so `ReptileMetaLearner.meta_update()` receives a well-formed `task_batch` without embedding sampling logic in the learner itself.

---

### `TaskSampler`

```python
class TaskSampler:
    """Samples TaskSample batches for meta-learning.
    
    Supports uniform, curriculum, and difficulty-aware strategies.
    Integrates with TaskRegistry for registered task routing.
    """
    
    def __init__(
        self,
        task_pool: Mapping[str, TaskProvider],
        strategy: SamplingStrategy = SamplingStrategy.UNIFORM,
        batch_size: int = 4,
        seed: int | None = None,
        metrics_registry: CollectorRegistry | None = None,
    ) -> None: ...
    
    async def sample_batch(self) -> list[TaskSample]: ...
    
    def update_difficulty(
        self,
        task_id: str,
        loss: float,
    ) -> None:
        """Update difficulty estimate for a task based on observed loss."""
        ...
    
    def register_task(
        self,
        task_id: str,
        provider: TaskProvider,
        difficulty: float = 0.5,
    ) -> None: ...
    
    def deregister_task(self, task_id: str) -> None: ...
```

### `SamplingStrategy` (enum)

```python
class SamplingStrategy(str, Enum):
    """Sampling strategy for task batches."""
    UNIFORM = "uniform"               # equal probability per task
    CURRICULUM = "curriculum"         # difficulty-sorted, anneal from easy
    DIFFICULTY_WEIGHTED = "difficulty_weighted"  # sample ∝ difficulty
    RECENCY_WEIGHTED = "recency_weighted"        # sample ∝ 1/(steps_since_sampled)
```

### `TaskProvider` (Protocol)

```python
class TaskProvider(Protocol):
    """Anything that can produce a TaskSample when called."""
    
    async def __call__(self, *, rng: np.random.Generator) -> TaskSample: ...
    
    @property
    def difficulty(self) -> float:
        """Difficulty estimate in [0, 1]. Updated by TaskSampler."""
        ...
```

### `CurriculumScheduler`

```python
class CurriculumScheduler:
    """Anneals task difficulty ceiling over meta-training steps.
    
    Implements the self-paced learning schedule:
        ceiling(t) = min(1.0, start_ceiling + slope * t)
    Only tasks with difficulty ≤ ceiling are eligible for sampling.
    """
    
    def __init__(
        self,
        start_ceiling: float = 0.3,
        slope: float = 0.01,
        warmup_steps: int = 100,
    ) -> None: ...
    
    def ceiling(self, step: int) -> float: ...
    
    def eligible_tasks(
        self,
        task_difficulties: Mapping[str, float],
        step: int,
    ) -> frozenset[str]: ...
```

### `SamplerSnapshot`

```python
@dataclass
class SamplerSnapshot:
    """Serialisable checkpoint of sampler state."""
    task_difficulties: dict[str, float]
    sample_counts:     dict[str, int]      # how many times each task sampled
    step:              int
    strategy:          SamplingStrategy
    timestamp:         float
```

---

## Integration with Phase 7.1

```
ReptileMetaLearner
    └─► meta_update(task_batch)
            ↑
        TaskSampler.sample_batch()     ← new in 7.2
            ↑
        TaskProvider (per task)
            ↑
        CurriculumScheduler.eligible_tasks()   ← filters by difficulty ceiling
```

Typical usage in the SLEEP_PHASE hook:

```python
async def _sleep_hook(self, step: int) -> None:
    batch = await self._task_sampler.sample_batch()
    result = await self._meta_learner.meta_update(batch)
    # Update difficulties based on per-task inner losses
    for task, loss in zip(batch, result.inner_losses):
        self._task_sampler.update_difficulty(task.task_id, loss)
```

---

## Prometheus Metrics

| Metric | Type | Description |
|---|---|---|
| `asi_task_sampler_batches_total` | Counter | Total batches sampled |
| `asi_task_sampler_tasks_sampled_total` | Counter (task_id label) | Per-task sample count |
| `asi_task_difficulty` | Gauge (task_id label) | Current difficulty estimate per task |
| `asi_curriculum_ceiling` | Gauge | Current difficulty ceiling |
| `asi_task_sampler_eligible_tasks` | Gauge | Number of eligible tasks at current ceiling |

---

## Test Targets (12)

1. `test_uniform_sampling_distributes_evenly` — chi-square on 1000 samples
2. `test_curriculum_blocks_high_difficulty_tasks` — ceiling=0.3 → hard task absent
3. `test_curriculum_unblocks_as_ceiling_rises` — step=500 → hard task reappears
4. `test_difficulty_weighted_favours_high_loss_tasks` — update_difficulty + sample distribution
5. `test_recency_weighted_avoids_recently_sampled` — recently-sampled task probability drops
6. `test_update_difficulty_clamps_to_unit_interval` — loss=99.9 → difficulty=1.0
7. `test_register_task_adds_to_pool` — register + sample_batch returns new task
8. `test_deregister_task_removes_from_pool` — deregister + sample_batch never returns it
9. `test_sample_batch_respects_batch_size` — len(batch) == batch_size always
10. `test_sampler_snapshot_round_trip` — snapshot + restore → identical distribution
11. `test_prometheus_metrics_pre_init` — all 5 counters/gauges present after __init__
12. `test_curriculum_scheduler_warmup` — step < warmup_steps → ceiling == start_ceiling

---

## Implementation Order

1. `SamplingStrategy` enum + `SamplerSnapshot` dataclass (`meta_types.py` additions)
2. `TaskProvider` Protocol + `CurriculumScheduler` (`task_provider.py`)
3. `TaskSampler.__init__` — pool registration, RNG seeding, Prometheus pre-init
4. `TaskSampler.sample_batch()` — UNIFORM and CURRICULUM strategies
5. `TaskSampler.sample_batch()` — DIFFICULTY_WEIGHTED and RECENCY_WEIGHTED strategies
6. `TaskSampler.update_difficulty()` — EMA smoothing (α=0.2 default)
7. `TaskSampler.register_task / deregister_task`
8. `CurriculumScheduler.eligible_tasks()` — difficulty ceiling filter
9. `SamplerSnapshot` serialise/restore
10. Tests — 12 targets across `test_task_sampler.py` and `test_curriculum.py`

---

## Labels

`phase-7`, `meta-learning`, `task-sampling`, `curriculum`

**Comment by web3guru888** [comment]
## Implementation Notes

### Module Placement

```
asi/
├── phase7/
│   ├── __init__.py          # add TaskSampler, SamplingStrategy, CurriculumScheduler
│   ├── meta_types.py        # add SamplingStrategy enum + SamplerSnapshot dataclass
│   ├── task_provider.py     # TaskProvider Protocol + CurriculumScheduler
│   └── task_sampler.py      # TaskSampler — core sampling logic
tests/
└── phase7/
    ├── conftest.py          # add sampler fixtures: mock_task_provider(), sampler_factory()
    ├── test_task_sampler.py # 10 of 12 targets
    └── test_curriculum.py   # 2 curriculum-specific targets (warmup + ceiling)
```

### `SamplingStrategy` enum — use `str` mixin

Use `class SamplingStrategy(str, Enum)` so YAML/JSON config files can pass the strategy as a plain string:

```python
class SamplingStrategy(str, Enum):
    UNIFORM = "uniform"
    CURRICULUM = "curriculum"
    DIFFICULTY_WEIGHTED = "difficulty_weighted"
    RECENCY_WEIGHTED = "recency_weighted"
```

`TaskSampler.__init__` accepts `strategy: SamplingStrategy | str` and coerces:
```python
self._strategy = SamplingStrategy(strategy)
```

### Difficulty EMA update

Don't store raw loss values — track an exponential moving average to smooth out noisy single-episode estimates:

```python
EMA_ALPHA = 0.2

def update_difficulty(self, task_id: str, loss: float) -> None:
    clamped = min(1.0, max(0.0, loss / self._loss_scale))  # normalise to [0,1]
    prev = self._difficulties.get(task_id, 0.5)
    self._difficulties[task_id] = EMA_ALPHA * clamped + (1 - EMA_ALPHA) * prev
    self._difficulty_gauge.labels(task_id=task_id).set(self._difficulties[task_id])
```

`loss_scale` (default=1.0) is a constructor kwarg — lets callers normalise by expected loss magnitude.

### CURRICULUM strategy — importance of `eligible_tasks`

When `ceiling < 1.0` and all tasks have difficulty > ceiling, `eligible_tasks()` returns an empty set. `sample_batch()` must fall back to the **easiest available task** rather than raising:

```python
eligible = self._scheduler.eligible_tasks(self._difficulties, self._step)
if not eligible:
    # fallback: pick the single easiest task repeated batch_size times
    easiest = min(self._difficulties, key=self._difficulties.get)
    eligible = frozenset([easiest])
```

### RECENCY_WEIGHTED — staleness counter

Maintain a `_steps_since_sampled: dict[str, int]` counter. Every call to `sample_batch()` increments all counters, then resets the counters for the tasks actually drawn:

```python
async def sample_batch(self) -> list[TaskSample]:
    for tid in self._steps_since_sampled:
        self._steps_since_sampled[tid] += 1
    # build weights for eligible tasks from recency counters
    weights_arr = softmax([self._steps_since_sampled[tid] for tid in eligible_list])
    chosen = self._rng.choice(eligible_list, size=batch_size, replace=True, p=weights_arr)
    for tid in chosen:
        self._steps_since_sampled[tid] = 0
```

### DIFFICULTY_WEIGHTED strategy

Difficulty-weighted sampling draws tasks proportional to their difficulty estimate (harder tasks sampled more often). This is complementary to CURRICULUM — CURRICULUM uses a ceiling to block tasks, DIFFICULTY_WEIGHTED uses raw proportional weights:

```python
weights = np.array([self._difficulties[tid] for tid in eligible_list])
weights = weights / weights.sum()   # normalise
chosen = self._rng.choice(eligible_list, size=batch_size, replace=True, p=weights)
```

### Prometheus cardinality guard

Same pattern as Phase 6.5 (`MultiTaskEWCRegulariser`): if the task pool exceeds 32 tasks, collapse label to `"other"` before emitting per-task metrics:

```python
_MAX_TASK_LABELS = 32

def _task_label(self, task_id: str) -> str:
    return task_id if task_id in self._top_tasks else "other"
```

Track `_top_tasks` as the set of the first 32 registered tasks.

### Test isolation — async sampler fixtures

All `TaskProvider` mocks must be `AsyncMock` (not `MagicMock`) since `__call__` is `async`:

```python
@pytest.fixture
def mock_provider() -> AsyncMock:
    m = AsyncMock(spec=TaskProvider)
    m.return_value = TaskSample(task_id="t0", support_set=..., query_set=..., labels=...)
    m.difficulty = 0.3
    return m
```

Chi-square uniformity test (target 1) should use `scipy.stats.chisquare` with `p > 0.01` threshold (not 0.05, to reduce flakiness).

### mypy Annotations

| Symbol | Type annotation |
|---|---|
| `task_pool` param | `Mapping[str, TaskProvider]` |
| `sample_batch` return | `list[TaskSample]` |
| `eligible_tasks` return | `frozenset[str]` |
| `update_difficulty` loss | `float` (not NDArray) |
| `_rng` | `np.random.Generator` (use `np.random.default_rng()`) |

**Comment by web3guru888** [comment]
## Phase 7 sub-phase tracker update

| Sub-phase | Feature | Issue | Status |
|---|---|---|---|
| 7.1 | Reptile meta-learning foundation | #259 | 🟡 Open |
| 7.2 | Task distribution sampler + curriculum | #262 | 🟡 Open |
| **7.3** | **Episodic Replay Buffer (PER)** | **#265** | **🟡 Open** |
| 7.4 | Hypernetwork context modulation | — | 📋 Planned |
| 7.5 | Phase 7 integration + benchmarks | — | 📋 Planned |

### Phase 7.3 filed: Prioritised Episodic Replay Buffer

Issue #265 covers `PrioritisedReplayBuffer` (sum-tree PER), `UniformReplayBuffer`, `Episode` dataclass, `ReplaySnapshot`, and the `build_replay_buffer()` factory. See discussions:
- #267 — Show & Tell: architecture, IS correction math, gradient blending
- #268 — Q&A: capacity, alpha, beta_start, update_priorities, Prometheus PromQL

SLEEP_PHASE hook order now has **8 steps** (buffer snapshot added as step 8).

**Comment by web3guru888** [comment]
## Phase 7.4 filed: Hypernetwork Context Modulation

**Issue #269** is now open: [Phase 7.4 — Hypernetwork Context Modulation](https://github.com/web3guru888/asi-build/issues/269)

### Phase 7 sub-phase tracker

| Sub-phase | Issue | Status |
|---|---|---|
| 7.1 Reptile Meta-Learning | #259 | 🟡 Open |
| 7.2 TaskSampler & Curriculum | #262 | 🟡 Open |
| 7.3 Episodic Replay Buffer | #265 | 🟡 Open |
| 7.4 Hypernetwork Modulation | #269 | 🟡 Open |
| 7.5 Consolidated Sleep-Phase | (upcoming) | ⬜ Planned |

### `SLEEP_PHASE` hook order (cumulative after 7.4)

1. `PrioritisedReplayBuffer.sample()` → replay batch
2. Replay gradient blend via `STDPOnlineLearner.update()`
3. `ReptileMetaLearner.meta_update(task_batch)`
4. `TaskSampler.sample_tasks()`
5. `HyperController.adapt(task_context)` ← **new in 7.4**
6. Online inference loop
7. `HyperController.restore_base()`
8. `FisherAccumulator.step()`
9. `EWCRegulariser.penalty_gradient()`

Phase 7.5 will wire all hooks into a single `SleepPhaseOrchestrator` with ordered execution and circuit-breaker fault tolerance.
